### PR TITLE
Port to .NET Core 3

### PR DESCRIPTION
**Pull request by 0xced**
Here's a pull request porting ICanHasDotnetCore to .NET Core 3 with a few side improvements. I have tried to break things in isolated commits so in case you don't want to merge it altogether you can cherry-pick commits of interest.

Here are the most important changes:
* Use Entity Framework Core instead of raw SQL
This change makes it super easy to develop with SQLite instead of SQL Server (I did all the work on macOS using JetBrains Rider). To develop on SQLite, just add this in `appsettings.Development.json`:
```json
{
  "Database": {
    "Provider": "Sqlite",
    "ConnectionString": "DataSource=/path/to/ICanHasDotnetCore.sqlite3"
  }
}
```

And maybe you could even use SQLite in production and save money on an SQL Server instance?

* Use [NuGet.Protocol](https://www.nuget.org/packages/NuGet.Protocol/) instead of NuGet.Core which is not compatible with .NET Core. This change has brought to light some packages that were incorrectly identified, the tests have been adapted accordingly.
* Improved logs overall. The logs added for Octokit allowed me to identify how to [improve performance of GitHub scanning](https://github.com/0xced/ICanHasDotnetCore/commit/689bb3d4b71de88ff263b1f89b8bef4f4423f55e).
* Web project modernization

Updated the web project to work with Node.js 12 and Gulp 4. I also added a new *Scan A NuGet package* tab to get the dependency graph of a NuGet package without having to upload a file.

And here are less important changes:

* Updated NuGet packages to their latest versions
* Replaced AutoFac with built-in ASP.NET Core equivalent features
* Replace Newtonsoft.Json with System.Text.Json
* Asyncified all code
* Updated knowledge
* Fixed a potential `ArgumentNullException`

### CODE REVIEW

File: source/Web/Features/Statistics/StatisticsRepository.cs [hunk 1]
```diff
@@ -1,125 +1,79 @@
...
-WHEN MATCHED THEN
-    UPDATE SET LatestSupportType = s.LatestSupportType, [Count] = [Count] + 1 
-
-WHEN NOT MATCHED THEN
-    INSERT (Name, LatestSupportType, Count)
-    VALUES (s.Name, s.LatestSupportType, 1);";
-
-            try
+            await using var context = _contextFactory();
+            var packageStatistic = await context.PackageStatistics.FindAsync(new object[] {package.PackageName}, cancellationToken);
```
Comment (droyad) [inline source/Web/Features/Statistics/StatisticsRepository.cs]: This is no longer atomic. I used the `MERGE` command as it guaranteed (from my understanding) atomic insert/update of that record. With this change two threads could update the same statistic record and only the last in wins.

**Comment by droyad** [comment]
@0xced I haven't forgotten about this and will get to doing a full review. I did a quick scan and only one thing stood out to me so far (added a comment)

File: source/Web/Features/Statistics/StatisticsRepository.cs (hunk 1 above)
Comment (0xced) [reply source/Web/Features/Statistics/StatisticsRepository.cs]: Good catch! I addressed this issue in 49561dba1e6f68497d69efe13669c1c8e0e0b6d1. I used [FlexLabs.EntityFrameworkCore.Upsert](https://github.com/artiomchi/FlexLabs.Upsert) as EF Core does not handle upsert out of the box.

By increasing the EF Core log level to information (`.MinimumLevel.Override("Microsoft.EntityFrameworkCore", LogEventLevel.Information)`) and enabling sensitive data logging (`optionsBuilder.EnableSensitiveDataLogging();`) we can see how FlexLabs.Upsert works:

With SQLite:

```sql
-- Executing DbCommand [Parameters=[@p0='Newtonsoft.Json' (Nullable = false) (Size = 15), @p1='1' (DbType = String), @p2='Supported' (Nullable = false) (Size = 9), @p3='1' (DbType = String)], CommandType='Text', CommandTimeout='30']
INSERT INTO "PackageStatistics" AS "T" ("Name", "Count", "LatestSupportType") VALUES (@p0, @p1, @p2) ON CONFLICT ("Name") DO UPDATE SET "Count" = ( "T"."Count" + @p3 )
```

With SQL Server:

```sql
-- Executing DbCommand [Parameters=[@p0='Newtonsoft.Json' (Nullable = false) (Size = 450), @p1='1', @p2='Supported' (Nullable = false) (Size = 4000), @p3='1'], CommandType='Text', CommandTimeout='30']
MERGE INTO [PackageStatistics] WITH (HOLDLOCK) AS [T] USING ( VALUES (@p0, @p1, @p2) ) AS [S] ([Name], [Count], [LatestSupportType]) ON [T].[Name] = [S].[Name] WHEN NOT MATCHED BY TARGET THEN INSERT ([Name], [Count], [LatestSupportType]) VALUES ([Name], [Count], [LatestSupportType]) WHEN MATCHED THEN UPDATE SET [Count] = ( [T].[Count] + @p3 );
```

**Comment by 0xced** [comment]
Excellent, I’ll stop force-pushing small changes while you are reviewing.

Would you consider using SQLite in production? I think for a simple two tables database (cache + stats) it would be well enough and cheaper than an SQL Server instance. Are you currently using an Azure SQL Database? Is it costly?

**Comment by 0xced** [comment]
I have pushed a few more improvements. We could probably shave a few more milliseconds when scanning GitHub repositories if https://github.com/octokit/octokit.net/pull/2151 gets merged and a new release of Octokit is released.

**Comment by droyad** [comment]
@0xced Thank you for this contribution. Unfortunately we've decided to shutter the site. The traffic has decreased significantly since .NET Core 3.1 was released (and previous 2.0). The ecosystem as a whole has transitioned over and the majority of use cases now work.

**Pull request closed without merging**